File: mercury/prime_factors.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: prime_factors.m
% Main author: nyc
%
% Small prime factorisation and related utilities.
%
%---------------------------------------------------------------------------%

:- module prime_factors.
:- interface.

:- import_module integer.
:- import_module list.
:- import_module positive.

%---------------------------------------------------------------------------%

    % factorise(N) = Factors:
    %
    % Factorise a positive integer into (prime, exponent) pairs,
    % with primes in ascending order.
    %
    % factorise(det_positive(integer(360))) =
    %   [{integer(2), 3}, {integer(3), 2}, {integer(5), 1}]
    %
:- func factorise(positive) = list(prime_power).

:- type prime_power
    --->    prime_power(
                pp_prime    :: integer,
                pp_exponent :: int
            ).

    % prime_factors(N) = Primes:
    %
    % Distinct prime factors of a positive integer, in ascending order.
    %
:- func prime_factors(positive) = list(integer).

    % is_prime(N):
    %
    % Trial-division primality test (adequate for the sizes we encounter).
    %
:- pred is_prime(integer::in) is semidet.

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- import_module int.
:- import_module require.

%---------------------------------------------------------------------------%

factorise(Pos) = Factors :-
    N = to_integer(Pos),
    ( if N = integer.one then
        Factors = []
    else
        trial_divide(N, integer(2), [], RevFactors),
        list.reverse(RevFactors, Factors)
    ).

    % trial_divide(N, P, !Acc):
    %
    % Divide out all powers of primes starting from P.
    %
:- pred trial_divide(integer::in, integer::in,
    list(prime_power)::in, list(prime_power)::out) is det.

trial_divide(N, P, !Acc) :-
    ( if N = integer.one then
        true
    else if P * P > N then
        % N is prime; record it.
        !:Acc = [prime_power(N, 1) | !.Acc]
    else
        count_factor(N, P, 0, Exp, Remainder),
        ( if Exp > 0 then
            !:Acc = [prime_power(P, Exp) | !.Acc],
            trial_divide(Remainder, next_candidate(P), !Acc)
        else
            trial_divide(N, next_candidate(P), !Acc)
        )
    ).

    % count_factor(N, P, !Exp, Remainder):
    %
    % Count how many times P divides N, returning the exponent
    % and the remainder after dividing out all copies of P.
    %
:- pred count_factor(integer::in, integer::in,
    int::in, int::out, integer::out) is det.

count_factor(N, P, !Exp, Remainder) :-
    integer.divide_with_rem(N, P, Q, R),
    ( if R = integer.zero then
        count_factor(Q, P, !.Exp + 1, !:Exp, Remainder)
    else
        Remainder = N,
        !:Exp = !.Exp
    ).

    % next_candidate(P) = NextP:
    %
    % After 2, try only odd numbers.
    %
:- func next_candidate(integer) = integer.

next_candidate(P) =
    ( if P = integer(2) then
        integer(3)
    else
        P + integer(2)
    ).

prime_factors(Pos) =
    list.map(func(prime_power(P, _)) = P, factorise(Pos)).

is_prime(N) :-
    N > integer.one,
    not has_small_factor(N, integer(2)).

:- pred has_small_factor(integer::in, integer::in) is semidet.

has_small_factor(N, P) :-
    P * P =< N,
    ( if integer.rem(N, P) = integer.zero then
        true
    else
        has_small_factor(N, next_candidate(P))
    ).

%---------------------------------------------------------------------------%
:- end_module prime_factors.
%---------------------------------------------------------------------------%

File: haskell/algebraic-polynomials/src/Math/Internal/PrimeFactors.hs
-- |
-- Module      : Math.Internal.PrimeFactors
-- Description : Small prime factorisation and related utilities
-- Stability   : experimental
--
-- Trial-division based prime factorisation and primality testing.
-- Adequate for the integer sizes encountered in polynomial and algebraic
-- number computations (typically up to a few thousand digits at most).
module Math.Internal.PrimeFactors
  ( factorise
  , primeFactors
  , isPrime
  , primes
  ) where

import Data.List.NonEmpty (NonEmpty(..))
import qualified Data.List.NonEmpty as NE
import Math.Internal.Positive (Positive, unPositive)

-- | Factorise a positive integer into @(prime, exponent)@ pairs,
-- in ascending order of prime.
--
-- Returns an empty list for the input @1@.
--
-- >>> factorise 360
-- [(2,3),(3,2),(5,1)]
-- >>> factorise 1
-- []
factorise :: Positive -> [(Integer, Int)]
factorise pos =
  let n = toInteger (unPositive pos)
  in if n == 1
     then []
     else map (\(x :| xs) -> (x, 1 + length xs)) . NE.group $ go n primes
  where
    go 1 _      = []
    go m (p:ps)
      | p * p > m = [m]
      | r == 0    = p : go q (p:ps)
      | otherwise = go m ps
      where (q, r) = m `divMod` p
    go _ []     = error "primes exhausted"  -- unreachable

-- | Distinct prime factors of a positive integer, in ascending order.
--
-- >>> primeFactors 360
-- [2,3,5]
primeFactors :: Positive -> [Integer]
primeFactors = map fst . factorise

-- | Trial-division primality test.
--
-- Checks divisibility by all primes up to \(\sqrt{n}\).
-- Adequate for the sizes encountered in this library.
--
-- >>> isPrime 7
-- True
-- >>> isPrime 6
-- False
isPrime :: Integer -> Bool
isPrime n
  | n < 2     = False
  | n < 4     = True
  | otherwise = all (\p -> n `mod` p /= 0) $ takeWhile (\p -> p * p <= n) primes

-- | Infinite list of prime numbers via trial division.
--
-- >>> take 5 primes
-- [2,3,5,7,11]
primes :: [Integer]
primes = 2 : filter isPrime [3, 5 ..]
